LD prune

tomwhite opened the pull request:
This is a first draft of #31, which is a port of (some of) the LD-prune code written by @eric-czech in https://github.com/related-sciences/gwas-analysis/tree/master/notebooks/platform/xarray/lib. It's not complete, but I thought I'd post my progress so far.

* The implementation uses the existing `windows` function in sgkit, so it only supports fixed size "units"-based indexing (number of variants, not base pair positions). However, when more flexible indexing is added (#341), then LD prune will be able to use that (hopefully with no changes).
* The core of the LD matrix calculation (`_ld_matrix_jit`) iterates over all the windows in a chunk, and within each window iterates over all the variant pairs in the window. The original implementation arranged things slightly differently, with each window broken up into multiple "axis intervals", which effectively unrolled the loop of the pairwise comparisons. The approach here makes it possible to decouple the array chunking from the window definition, so you can use the existing Dask array chunks (the original implementation calculates its own chunks).
* The relevant unit tests have been ported, including a concordance check with scikit-allel's `locate_unlinked` function. This uses Hypothesis to generate lots of combinations of input array, windows, and chunk sizes. To get an exact match, I had to declare the numba types for the Rogers Huff R function to be single-precision floats. (There are slight differences if they are left as the default double precision.)
* The biggest outstanding issue is that the `ld_matrix` function returns a Dask dataframe, rather than an Xarray Dataset - which would be more consistent with the rest of sgkit. It would also be better to provide `scores` as a Dataset variable, rather than an array, since its dimension could then be validated.
* Documentation is yet to be added. As a part of this PR I will add function docs (and expose `ld_matrix`, `maximal_independent_set`, and `ld_prune` in the public API).

jeromekelleher commented:
Looks great @tomwhite. One quick thought: I wonder if we could pull out a PR to just compute LD (r2) first? We'll definitely want this and it would reduce the size of this PR a bit.

tomwhite commented:
@jeromekelleher good suggestion. I'll need to write some unit tests for that function, which will be a good thing.

tomwhite commented:
I've rebased this to include #536, and I've added a way to optionally specify scores as a dataset variable.

I had a quick look at having `ld_matrix` return an Xarray Dataset rather than a Dask DataFrame, but I think that the computation would be eager, since the number of rows in the result is unknown (there is a bit of discussion about it here: https://github.com/pydata/xarray/issues/3929).

It also wasn't particularly obvious what the variable names would be: `ld_variant_0` and `ld_variant_1` for `i` and `j`, perhaps, but `maximal_independent_set` is not just for LD calculations, so these might not be quite right for that.

So I'm not sure what to do about this - we could perhaps not expose `ld_matrix` or `maximal_independent_set` as public methods - or we could have them as public utility methods that use Dask DataFrames.

tomwhite commented:
* As discussed in the [meeting](https://github.com/pystatgen/sgkit/discussions/553), I've left the return value of `ld_matrix` and the input to `maximal_independent_set` as a Dask dataframe.
* I've also left the variables in the dataframe as `i`, `j` etc, since they don't form part of a larger dataset, so they an be considered "local" variables. Happy to change them if there were some better names for them.
* I've changed `ld_prune` to return a dataset where the variants in linkage disequilibrium have been removed. This is similar to what Hail does. For more control over the process, it's straightforward to use `ld_matrix` followed by `maximal_independent_set` to find the variants to drop.
* I've added documentation and an example.

This still depends on #536.
